### DISCUSSION TITLE
500 on mutation without meaningful error message

### DISCUSSION THREAD

**Discussion by antoinelin** [urql-graphql/urql#1739] [Q&A] [answered]
Hey 🖖🏻 

I would like to trigger a mutation using Urql, but I can't make my request working. I'm always getting a 500 with the following message: `POST body missing. Did you forget use body-parser middleware?`

I followed the documentation to create my mutation and my query+server are working fine by using Postman or Playground. Also, I tried to trigger the mutation using `graphql-request` instead of `Urql` and it worked as well (same query, same fetchOptions).

Any idea on where I should dig to find a solution ?

Thanks 💪🏻

**Comment by kitten** [comment, marked as answer]
so, the answer is indeed right in front of you in that error 😅 Your API is receiving the POST request but is not parsing the JSON body, so you'll need to use the mentioned middleware for it to be parsed 

it's likely that it's not quite working because there's some difference in what you're doing with graphql-request, like a different endpoint, GET rather than POST, etc

**Reply by antoinelin** [reply]
Thank you for your answer but the mentioned middleware is existing on the server. What I was meaning by "without meaningful error message" is that I don't think that the issue is the body parsing of the request (did not chose the right word I guess, not an english native speaker :( ).

There is not differences between my tests by using `graphql-request` and `urql`. And the POST query is working well by using `graphql-request`, `postman` and the GraphQL playground:

With `urql`:
<img width="363" alt="Capture d’écran 2021-06-23 à 07 42 26" src="https://user-images.githubusercontent.com/11920757/123042744-8cf23b00-d3f7-11eb-80eb-cb0eae2fa563.png">

With `graphql-request`:
<img width="346" alt="Capture d’écran 2021-06-23 à 07 45 16" src="https://user-images.githubusercontent.com/11920757/123042815-a8f5dc80-d3f7-11eb-89cf-6d4694f8f73f.png">

With `postman`:
<img width="1047" alt="Capture d’écran 2021-06-23 à 07 50 56" src="https://user-images.githubusercontent.com/11920757/123042927-d80c4e00-d3f7-11eb-8293-b9d18bcfff5b.png">

I tried by using exactly the same payload as provided to the request by `urql` on Postman and it works well. The only difference between `urql` and `graphql-request` request is that `urql` is adding `operationName` to the payload which is missing in `graphql-request`.

For all the queries my request headers are the same (I use `content-type: application/json` everywhere), CORS is working well and I use POST method. I don't use authorization on localhost.
Server side, I use Koa with `koa-body` and `koa-json`. I tried `koa-bodyparser` without success neither. As the request `content-type` header is `application/json` I should not have to use `koa-bodyparser-graphql`.

If it was a JSON body parsing issue, I think it would not worked anywhere am I wrong (by using exactly same request parameters I mean) ?

**Reply by antoinelin** [reply]
Update: I succeed to make it works by using `client.mutation` back-end side (api routes on Next.js) with the same client as the front-end one. But this is not an ideal solution imo

**Reply by JoviDeCroock** [reply]
I mean, ideally a reproduction would be really neat 😅 

- [Client template](https://codesandbox.io/s/urql-issue-template-client-iui0o)
- [Server template](https://codesandbox.io/s/urql-issue-template-server-0ufyz)

Because this honestly sounds like a missing thing or a bug in how the body is parsed `operationName` should not be an issue as that is an established convention. In this example `apollo-server` is used but we've got different server implementations that also work 😅 we can only troubleshoot a limited amount over pure text without seeing an example

**Reply by antoinelin** [reply]
Sure, understand this sounds kind of weirdly (it was for me), will try to give a repro asap! Was trying if someone else had the same issue (that's why I first created a discussion instead of an issue). Thanks for the answer :)
